zipline/data/hdf5_daily_bars.py

```diff
@@ -0,0 +1,758 @@
+        /end_date
+    /CA
+      /data
+        /open
+        /high
+        /low
+        /close
+        /volume
+      /index
+        /sid
+        /day
+      /lifetimes
+        /start_date
+        /end_date
+"""
+
+
+from functools import partial
+
+import h5py
+import logbook
+import numpy as np
+import pandas as pd
+from six import iteritems
+from six.moves import reduce
+
+from zipline.data.bar_reader import NoDataBeforeDate, NoDataAfterDate
+from zipline.data.session_bars import SessionBarReader
+from zipline.utils.memoize import lazyval
+
+
+log = logbook.Logger('HDF5DailyBars')
+
+
+DATA = 'data'
+INDEX = 'index'
+LIFETIMES = 'lifetimes'
+
+SCALING_FACTOR = 'scaling_factor'
+
+OPEN = 'open'
+HIGH = 'high'
+LOW = 'low'
+CLOSE = 'close'
+VOLUME = 'volume'
+
+FIELDS = (OPEN, HIGH, LOW, CLOSE, VOLUME)
+
+DAY = 'day'
+SID = 'sid'
+
+START_DATE = 'start_date'
+END_DATE = 'end_date'
+
+
+DEFAULT_SCALING_FACTORS = {
+    # Retain 3 decimal places for prices.
+    OPEN: 1000,
+    HIGH: 1000,
+    LOW: 1000,
+    CLOSE: 1000,
+    # Volume is expected to be a whole integer.
+    VOLUME: 1,
+}
+
+
+def coerce_to_uint32(a, field, scaling_factor):
+    """
+    Returns a copy of the array as uint32, applying a scaling factor to
+    maintain precision if supplied.
+    """
+    return (a * scaling_factor).astype('uint32')
+
+
+def days_and_sids_for_frames(frames):
+    """
+    Returns the date index and sid columns shared by a list of dataframes,
+    ensuring they all match.
+
+    Parameters
+    ----------
+    frames : list[pd.DataFrame]
+        A list of dataframes indexed by day, with a column per sid.
+
+    Returns
+    -------
+    days : np.array[datetime64[ns]]
+        The days in these dataframes.
+    sids : np.array[int64]
+        The sids in these dataframes.
+
+    Raises
+    ------
+    ValueError
+        If the dataframes passed are not all indexed by the same days
+        and sids.
+    """
+    frame_0 = frames[0]
+
+    # Ensure the indices and columns all match.
+    for frame_1 in frames[1:]:
+        if not frame_0.index.equals(frame_1.index):
+            raise ValueError('Frames have mistmatched days.')
+
+        if not frame_0.columns.equals(frame_1.columns):
+            raise ValueError('Frames have mismatched sids.')
+
+    return frame_0.index.values, frame_0.columns.values
+
+
+class HDF5DailyBarWriter(object):
+    """
+    Class capable of writing daily OHLCV data to disk in a format that
+    can be read efficiently by HDF5DailyBarReader.
+
+    Parameters
+    ----------
+    filename : str
+        The location at which we should write our output.
+    date_chunk_size : int
+        The number of days per chunk in the HDF5 file. If this is
+        greater than the number of days in the data, the chunksize will
+        match the actual number of days.
+
+    See Also
+    --------
+    zipline.data.hdf5_daily_bars.HDF5DailyBarReader
+    """
+    def __init__(self, filename, date_chunk_size):
```

Comment:
One thing we might want to consider here is having this take a fully-formed object rather than just a file path. Doing so would allow us to do things like write to an in-memory h5 file, which we can't express in the current design. We could still have this take a filepath and use `coerce_types` (or just code in the constructor) to convert to the relevant object (this is how AssetFinder works, for example). Is there a reason that we want/need this to defer creation until a `write` call? One issue might be that we'd probably want to make this a context manager to ensure that the file gets closed, which would be a fairly different API from the rest of our writers.

Reply:
My instinct is probably to leave this as-is for now given the worry about inconsistency with other writers. I could be convinced otherwise though.

Reply:
I actually came to this PR to make this same request because of how I want to use this in our writer service. If you want to make a reader over the recently written object, there is currently a risk of opening two handles to the same h5 file which is not totally safe.